src/lib.rs

```diff
@@ -2312,13 +2325,50 @@ impl Connection {
             left > frame::MAX_CRYPTO_OVERHEAD &&
             !is_closing
         {
-            let crypto_len = left - frame::MAX_CRYPTO_OVERHEAD;
-            let crypto_buf = self.pkt_num_spaces[epoch]
+            let max_len = left - frame::MAX_CRYPTO_OVERHEAD;
+
+            let crypto_off =
+                self.pkt_num_spaces[epoch].crypto_stream.send.off_front();
+
+            // Encode the frame.
+            //
+            // Instead of creating a `frame::Frame` object, encode the frame
+            // directly into the packet buffer.
+            //
+            // First we reserve some space in the output buffer for writing the
+            // frame header (we assume the length field is always a 2-byte
+            // varint as we don't know the value yet).
+            //
+            // Then we emit the data from the crypto stream's send buffer.
+            //
+            // Finally we go back and encode the frame header with the now
+            // available information.
+            let hdr_off = b.off();
+            let hdr_len = 1 + // frame type
+                octets::varint_len(crypto_off) + // offset
+                2; // length, always encode as 2-byte varint
+
+            let (mut crypto_hdr, mut crypto_payload) =
+                b.split_at(hdr_off + hdr_len)?;
+
+            // Write stream data into the packet buffer.
+            let (len, _) = self.pkt_num_spaces[epoch]
                 .crypto_stream
                 .send
-                .pop(crypto_len)?;
+                .emit(&mut crypto_payload.as_mut()[..max_len])?;
+
+            // Encode the frame's header.
+            crypto_hdr.skip(hdr_off)?;
```

Comment:
This took a couple of read-overs. I'm still a bit unclear why the skip is required. Is it because `crypto_hdr` contains something before the space for the crypto header?

Reply:
The `split_at()` method is kind of weird, because it splits the buffer from its beginning, rather than the current offset (this IIRC is required for the packet encryption/decryption). So `crypto_hdr` will actually start at the very beginning of the packet and needs to be advanced to the actual offset of the frame. I can probably add a comment to clarify.

Reply:
yeah a comment or a slight rename of the variable would be fine.

Reply:
I added some more commentary, hopefully it improves things.